metaci/release/tasks.py

```diff
@@ -1,16 +1,149 @@
+from metaci.plan.models import PlanRepository
 from metaci.release.models import Release, ReleaseCohort
 from metaci.repository.models import Repository
 
-from github3.repos.repo import Repository as GitHubRepository
+
+def _run_planrepo_for_release(release: Release, planrepo: PlanRepository):
+    build = Build(
+        repo=release.repo,
+        plan=planrepo.plan,
+        planrepo=planrepo,
+        commit=release.created_from_commit,
+        build_type="auto",
+        release=release,
+        release_relationship_type="automation",
+    )
+    build.save()
+    if release.status != Release.STATUS.inprogress:
+        release.status = Release.STATUS.inprogress
+        release.save()
+
+
+def find_active_planrepos_by_role(
+    release: Release, role: str
+) -> QuerySet[PlanRepository]:
+    return release.repo.planrepos.filter(plan__role=role, active=True)
+
+
+def fail_release_with_message(release: Release, message: str):
+    release.error_message = _(message)
+    release.status = Release.STATUS.failed
+    release.save()
+
+    release.release_cohort.error_message = _("One or more releases failed.")
+    release.release_cohort.status = Release.STATUS.failed
+    release.release_cohort.save()
+
+
+def _run_release_builds(release: Release):
+    FAILED_STATUSES = [BUILD_STATUSES.error, BUILD_STATUSES.fail]
+    COMPLETED_STATUSES = [BUILD_STATUSES.success, *FAILED_STATUSES]
+
+    # NOTE: bug where MetaCI builds can hang in In Progress forever
+    # will be painful here.
+    def running(builds: List[Build]) -> bool:
+        return bool(builds) and not all(b.status in COMPLETED_STATUSES for b in builds)
+
+    def last_succeeded(builds: List[Build]) -> bool:
+        return (
+            bool(builds)
+            and not running(builds)
+            and builds[-1].status == BUILD_STATUSES.success
+        )
+
+    def last_failed(builds: List[Build]) -> bool:
+        return (
+            bool(builds)
+            and not running(builds)
+            and builds[-1].status in FAILED_STATUSES
+        )
+
+    def any_failed(builds: List[Build]) -> bool:
+        return bool(builds) and any(b.status in FAILED_STATUSES for b in builds)
+
+    if release.status in [Release.STATUS.failed, Release.STATUS.completed]:
+        # Release manager must manually set the status back to In Progress if they're attempting
+        # to retry any failed builds.
+        return
+
+    # Locate builds associated with this Release.
+    # Sort Builds so that we can find the latest of each Role.
+    # This supports multi-build retries (rather than rerunning a single Build).
+    # Note that our predicates (succeeded() et al.) check for running builds
+    # and return False if builds are still running, which is why we can
+    # reasonably use `time_end`.
+
+    builds = defaultdict(list)
+    for build in release.build_set.filter(
+        plan__role__in=["Upload Release", "Release Deploy", "Release Test"]
```

Comment:
Need to use the internal role names: https://github.com/SFDO-Tooling/MetaCI/blob/main/metaci/plan/models.py#L21

Reply:
(And similarly below)